src/app/elevation/qgselevationprofilelayertreemodel.cpp

```diff
@@ -0,0 +1,246 @@
+/***************************************************************************
+                          qgselevationprofilelayertreemodel.cpp
+                          -----------------
+    begin                : April 2022
+    copyright            : (C) 2022 by Nyall Dawson
+    email                : nyall dot dawson at gmail dot com
+***************************************************************************/
+
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgselevationprofilelayertreemodel.h"
+#include "qgslayertreenode.h"
+#include "qgslayertree.h"
+#include "qgssymbollayerutils.h"
+#include "qgsvectorlayerelevationproperties.h"
+#include "qgsmeshlayerelevationproperties.h"
+#include "qgsrasterlayerelevationproperties.h"
+#include "qgsvectorlayer.h"
+#include "qgssinglesymbolrenderer.h"
+#include "qgsmarkersymbol.h"
+#include "qgsfillsymbol.h"
+
+QgsElevationProfileLayerTreeModel::QgsElevationProfileLayerTreeModel( QgsLayerTree *rootNode, QObject *parent )
+  : QgsLayerTreeModel( rootNode, parent )
+{
+  setFlag( QgsLayerTreeModel::AllowNodeReorder );
+  setFlag( QgsLayerTreeModel::AllowNodeChangeVisibility );
+  setFlag( QgsLayerTreeModel::ShowLegendAsTree );
+  setFlag( QgsLayerTreeModel::AllowLegendChangeState, false );
+}
+
+QVariant QgsElevationProfileLayerTreeModel::data( const QModelIndex &index, int role ) const
+{
+  switch ( role )
+  {
+    case Qt::DecorationRole:
```

Comment:
Just wondering if it wouldn't be more flexible to have a virtual method for elevation properties classes, which would return QgsLayerTreeModelLegendNode (or a list of them) - it could keep the layer-related logic more contained... (but could be a complication as well)

Reply:
>  (but could be a complication as well)

At this stage I'm leaning toward it being an added complication -- we'd then need to override a bunch of things in the layer tree for populating the legend nodes too. If this logic gets more complex then I'll revisit in future :+1: